### ISSUE TITLE
Fields labeled t.uuid in schema fail to return a UUIDTools::UUID instance in Rails 5.1

### ISSUE THREAD

**Issue by kevinwongso** [riboseinc/activeuuid#1]
In Rails 5.1, fields labeled `t.uuid` or the `id` field on classes that have `include ActiveUUID::UUID` no longer return a `UUIDTools::UUID` instance, instead returning a string encoded as "`ASCII-8BIT`", which corresponds to calling `.raw` on a `UUIDTools::UUID` instance returned by the regular behavior.

E.g. in Rails 5.1
```
MyModel.first.id
=> "$A\xBBa\xA5^MO\xB1\r\xAC\x95e\x12\x8A\xE6"
```

Old behavior in Rails 5.0 and below
```
MyModel.first.id
=> #<UUID:0x2b0cfcee945c UUID:2441bb61-a55e-4d4f-b10d-ac9565128ae6>
MyModel.first.id.raw
=> "$A\xBBa\xA5^MO\xB1\r\xAC\x95e\x12\x8A\xE6"
```

**Comment by ronaldtse** [comment]
@skalee can you help fix this? Thanks!

**Comment by ronaldtse** [comment]
Ping @skalee 👍  I think @ribose-jeffreylau 's work is in the "rails5-wip" branch.

We need to bring Rails 5.1 compatibility to this gem, and perhaps we can somehow meld this with the Rails 5.1 approach of UUID -- seems that the only difference is that Rails uses `string` for UUID but we use a `binary(16)` field. This involves:

* Dumping of the `db/schema.rb`
* Models should provide an ID that is a UUID model read from whatever "UUIID representation" stored in the datastore (which could be `.to_s` to a String)

**Comment by kevinwongso** [comment]
This might be relevant:
https://blog.metova.com/rails-5-attributes-api

And this gem seems to be an example of someone's attempt to use Rails 5 Attributes API to achieve something similar to ActiveUUID:
https://github.com/nedap/mysql-binuuid-rails
Though, according to a test case someone added, the gem still can't handle conversion from complex queries due to how Attributes API works, so pure Attributes API may not be enough.

**Comment by ronaldtse** [comment]
@kevinwongso doesn't this mean we can just use the binuuid gem and get rid of activeuuid? Can you try?

**Comment by kevinwongso** [comment]
I'll try checking what the gem is trying to do, and see if we can apply it to activeuuid or if it's enough to just use the gem to replace activeuuid. I'm concerned about the failed test case discussed on that gem as our own projects that use activeuuid have non-trivial SQL queries that may fail similarly.

**Comment by ronaldtse** [comment]
Right, but it might be easier to fix that approach than to redo this. Let’s see what happens!

**Comment by kevinwongso** [comment]
[#cfad7646caed5732a73c2166886af0ab010eb993](https://github.com/riboseinc/activeuuid/commit/cfad7646caed5732a73c2166886af0ab010eb993)
Seems to fix it for Rails 5.1, issue seems to be that the Patches module that ActiveUUID prepends to register the types had an if statement that checks for Rails major & minor versions. It wasn't updated for 5.1, so there was a step that is only executed if Rails major version is 5 and the minor version is 0. Running it in a Rails 5.1 environment thus skips that step. I changed the if statement to run that step in 5.1 too, and added a "cast" method to the UUID type that 5.1 seems to look for.

After these changes, the models seem to work fine again and properly return a UUIDTools::UUID instance when `.id` is invoked.
